### ISSUE TITLE
Handle additional fields in ClusterClass variables

### ISSUE THREAD

**Issue by sbueringer** [kubernetes-sigs/cluster-api#6135]
**User Story**

As a user I would like to get an error when I specify additional (undefined) fields in variables.

**Detailed Description**

Let's assume you have defined a ClusterClass variable which is an object (or array of objects, ...). If users now provide variable values on the Cluster resource with additional fields they are stored in etcd and then later used in patches. This can lead to continuously created objects (as shown by #6113). The topology reconciler is unable to create the templates with those fields as they are just pruned by the Kubernetes apiserver.

It would be great if we can just return an error on Cluster create/update in those cases.

**Prio Art**:

In upstream Kubernetes there are two features for this:
1. kubectl on client-side validates the OpenAPI schema and returns an error like this:
  > error: error validating "./cluster-bug.yaml": error validating data: ValidationError(Cluster.spec.topology): unknown field "abc" in io.x-k8s.cluster.v1beta1.Cluster.spec.topology; if you choose to ignore these errors, turn validation off with --validate=false
2. If validate=false is used the APIserver prunes unknown fields, before storing the object in etcd. I think this should be relatively easy to do by calling the Prune upstream func in our mutating ("defaulting") webhook.

**Suggestion**

I think we have to take a closer look at the Kubernetes implementation. Ideally we implement 1. in our webhook, if that is not possible / or as a safeguard 2 can be implemented as well.

**Additional notes**

I think the same could apply to ClusterClass variable defaults/enums/examples (but I have to check that)

Extracted from: https://github.com/kubernetes-sigs/cluster-api/issues/6113#issuecomment-1040043990

/kind feature
/area topology

**Comment by imikushin** [comment]
Perhaps, if `type: object` is used without further specification (defining the object fields), it could be viewed as "free form" object. This way existing clusters will not be affected.

The conversion webhook to a later version of Cluster API (introducing stricter validation) can add `x-kubernetes-preserve-unknown-fields: true` to such `type: object` variables to explicitly mark them as "free form".

**Comment by sbueringer** [comment]
I'm not 100% sure we can use the conversion webhook to migrate resources in this way without hitting weird edge cases.

**Comment by fabriziopandini** [comment]
+1 to take steps to make opting in into free form objects explicit and adding pruning; I think this is a must-have before graduating ClusterClass to beta because users might be tempted to abuse free form objects, but lack of schema can lead to a much bigger problem down the line like e.g. infinite reconcile (we already had issue reported about this).
I will send a PR to add a warning in the documentation

Let's discuss the idea of the conversion webhook + a phased approach in the office hours.
/priority important-soon

**Comment by sbueringer** [comment]
Agree, although I'm not sure what graduating to beta means technically at this point. I assume enabling the ClusterTopology feature flag per default? (as the API is already v1beta1)

**Comment by sbueringer** [comment]
Okay so here's an overview.

Overall we want to have the following behavior eventually:
* folks can opt into free form object variables in the variable schema via `x-kubernetes-preserve-unknown-fields: true` (same as in CRDs)
* if `x-kubernetes-preserve-unknown-fields: true` is not set and there are additional fields in a variable object value we return an error in the Cluster validation webhook 
    * Note: Yuvaraj and I investigated that a while back and the `Prune` func in the upstream library makes it easy to detect if there are additional fields. We would prefer just returning an explicit error in that case over implicitly pruning. The first is way less surprising for users.

Migration options:

1. Introducing `x-kubernetes-preserve-unknown-fields` and validation for additional fields with v1.2
2. Introducing `x-kubernetes-preserve-unknown-fields` with v1.2
  a.) Introducing validation for additional fields with v1.3
  b.) Introducing validation for additional fields with v1beta2

**Comment by neolit123** [comment]
does `x-kubernetes-preserve-unknown-fields` catch case-sensitively unknown fields - e.g. `foo` vs `Foo`?
IIRC, this was blocked by json parsers in the past, but might have changed after https://sigs.k8s.io/json was created.

**Comment by sbueringer** [comment]
Hm. I'm not sure we would have to try. I would like to aim for a behavior which is consistent with Kubernetes CRD behavior given that we use their packages to do the actual validation (detail: we would call their Prune func in our validation webhook and the Prune func tells us if it pruned additional fields).

**Comment by sbueringer** [comment]
> The topology reconciler is unable to create the templates with those fields as they are just pruned by the Kubernetes apiserver on object creation.

It looks like this is not true anymore with SSA. Apparently SSA detects this case:
> failed to create typed patch object (default/my-cluster-jgrq7; [infrastructure.cluster.x-k8s.io/v1beta1](http://infrastructure.cluster.x-k8s.io/v1beta1), Kind=DockerCluster): .spec.subnets: field not declared in schema

(but let's please verify this to be 100% sure)

**Comment by sbueringer** [comment]
> > The topology reconciler is unable to create the templates with those fields as they are just pruned by the Kubernetes apiserver on object creation.
> 
> It looks like this is not true anymore with SSA. Apparently SSA detects this case:
> 
> > failed to create typed patch object (default/my-cluster-jgrq7; [infrastructure.cluster.x-k8s.io/v1beta1](http://infrastructure.cluster.x-k8s.io/v1beta1), Kind=DockerCluster): .spec.subnets: field not declared in schema
> 
> (but let's please verify this to be 100% sure)

Verified:

ClusterClass:
```yaml
  - name: lbImage
    required: true
    schema:
      openAPIV3Schema:
        type: object
        properties:
          imageRepository:
            type: string
            default: kindest
          imageTag:
            type: string
            default: v20210715-a6da3463
```
+ Cluster:
```yaml
    - name: lbImage
      value:
        imageRepository: kindest
        imageTag: v20210715-a6da3463
        abc: def
```

=> logs 
> {"ts":1664297292587.2688,"caller":"controller/controller.go:326","msg":"Reconciler error","controller":"topology/cluster","controllerGroup":"cluster.x-k8s.io","controllerKind":"Cluster","Cluster":{"name":"my-cluster","namespace":"default"},"namespace":"default","name":"my-cluster","reconcileID":"2d2d0738-81c2-436d-96b1-3a303d4ae7ee","err":"error reconciling the Cluster topology: failed to create patch helper for DockerCluster/my-cluster-mzv79: failed to request dry-run server side apply: failed to create typed patch object (default/my-cluster-mzv79; infrastructure.cluster.x-k8s.io/v1beta1, Kind=DockerCluster): .spec.loadBalancer.abc: field not declared in schema","errVerbose":"failed to create typed patch object (default/my-cluster-mzv79; infrastructure.cluster.x-k8s.io/v1beta1, Kind=DockerCluster): .spec.loadBalancer.abc: field not declared in schema\nfailed to request dry-run server side apply\nsigs.k8s.io/cluster-

This means we shouldn't run into infinite reconciles because of this because SSA errors for unknown fields.

But this still leaves the case were folks are not specifying the entire object (either no fields at all or not all of them) and depend on them being there.

So I think we should still try to find a way to introduce `x-kubernetes-preserve-unknown-fields: true` and return errors in cases where additional fields are present via the webhook.

I wonder if it's fine to just introduce this behavior in a new Cluster API version.

Essentially with that new Cluster API version we would return errors on create/update on Cluster CRs if undefined fields are present and `x-kubernetes-preserve-unknown-fields` is **not** set.

This wouldn't block the reconcile it just means that for all future Cluster create/updates there has to be an explicit opt-in into unknown fields or they cannot be present. This can then be fixed by users by either updating the ClusterClass or the Cluster accordingly (depending on if the additional field is actually desired or not)

**Comment by sbueringer** [comment]
Thought about it a bit more & discussed it with Fabrizio, to summarize:

* with v1.3 we would make the Cluster & ClusterClass validation webhooks more strict:
  * if `x-kubernetes-preserve-unknown-fields` is set: additional fields are allowed
  * otherwise: validation webhooks fail if additional fields are found
  * We would allow to opt-out from the stricter validation if the `unsafe.topology.cluster.x-k8s.io/disable-strict-field-validation` annotation is present on the ClusterClass
    * The annotation would be immediately deprecated, so we can drop it with v1beta2.

Please note that the Cluster topology reconciliation is not affected.

If users then get the error that not all fields are defined in the schema they can:
* update the variable schema in the ClusterClass to add the fields
* update the variable in the Cluster to drop the additional fields
* opt-out from the validation with the annotation on the ClusterClass

**Comment by sbueringer** [comment]
I have some free time to implement this

/unassign @ykakarap 
/assign @sbueringer

**Comment by fabriziopandini** [comment]
Given that users are required to edit the CC for adding the annotation, at this point I think that we should ask them to add x-Kubernetes-preserve-unknown-fields: true instead which is the permanent way to opt out (while the annotation is just temporary)

In other words, ok for me but let's keep things simple and not implement the temporary annotation
